src/seabreeze/pyseabreeze/transport.py

```diff
@@ -364,3 +368,254 @@ def get_name_from_pyusb_backend(backend: usb.backend.IBackend) -> str | None:
+            raise TypeError("device needs to be an IPv4TransportHandle")
+        # TODO handle possible exceptions
+        self._device = device
+        self._opened = True
+
+        # This will initialize the communication protocol
+        if self._opened:
+            self._protocol = self._protocol_cls(self)
+
+    @property
+    def is_open(self) -> bool:
+        return self._opened or False
+
+    def close_device(self) -> None:
+        if self._device is not None:
+            self._device.close()
+            self._device = None
+        self._opened = False
+        self._protocol = None
+
+    def write(self, data: bytes, timeout_ms: int | None = None, **kwargs: Any) -> int:
+        if self._device is None:
+            raise RuntimeError("device not opened")
+        if kwargs:
+            warnings.warn(f"kwargs provided but ignored: {kwargs}")
+        if timeout_ms:
+            self._device.socket.settimeout(timeout_ms / 1000.0)
+        return self._device.socket.send(data)
+
+    def read(
+        self,
+        size: int | None = None,
+        timeout_ms: int | None = None,
+        mode: str | None = None,
+        **kwargs: Any,
+    ) -> bytes:
+        if self._device is None:
+            raise RuntimeError("device not opened")
+        if size is None:
+            # use minimum packet size (no payload)
+            size = 64
+        if kwargs:
+            warnings.warn(f"kwargs provided but ignored: {kwargs}")
+        if timeout_ms:
+            self._device.socket.settimeout(timeout_ms / 1000.0)
+        data = bytearray(size)
+        toread = size
+        view = memoryview(data)
+        while toread:
+            nbytes = self._device.socket.recv_into(view, toread)
+            view = view[nbytes:]
+            toread -= nbytes
+        return bytes(data)
+
+    @property
+    def default_timeout_ms(self) -> int:
+        if not self._device:
+            raise RuntimeError("no protocol instance available")
+        timeout = self._device.socket.gettimeout()
+        if not timeout:
+            return 10000
+        return int(timeout * 1000)
+
+    @property
+    def protocol(self) -> PySeaBreezeProtocol:
+        if self._protocol is None:
+            raise RuntimeError("no protocol instance available")
+        return self._protocol
+
+    @classmethod
+    def list_devices(cls, **kwargs: Any) -> Iterable[IPv4TransportHandle]:
+        """list IPv4 devices for all available spectrometers
+
+        Note: this includes spectrometers that are currently opened in other
+        processes on the machine.
+
+        Yields
+        ------
+        devices : IPv4TransportHandle
+            unique socket devices for each available spectrometer
+        """
+        dev_sockets = []
+
+        # Use multicast to discover potential spectrometers. Requires a network
+        # adapter to be specified.
+        network_adapter = kwargs.get("network_adapter", None)
```

Comment:
It would be great if we could fallback to just use all discoverable network adapters.
A user might not want to think about this explicitly.

Reply:
Yes, I was thinking the same and have modified this to use `socket.IFADDR_ANY` if no network adapter has been specified. *However*, this does not mean that the request will be sent to all interfaces (unfortunately): only to a single one, and on a Linux system, that is the one with the highest metric. It is a sensible default, but to send on all interfaces, one would have to determine these and manually send the request to each one. There is, to my knowledge, no built-in way of doing so: one would need to use e.g. `psutil` to list all available interfaces and loop over them.

I think the solution with `IFADDR_ANY` is what most users would expect and works fine with a single interface.